pkg/providers/docker/reconciler/reconciler_test.go

```diff
@@ -203,6 +278,49 @@ func (tt *reconcilerTest) allObjs() []client.Object {
 	return objs
 }
 
+func (tt *reconcilerTest) runReconcileControlPlaneTestFlow() {
```

Comment:
I understand the idea here, however I would vote against this. It does avoid repeated code which is good, however, drying out tests too much can make them more difficult to read. This almost creates its own framework for tests.

It you still think this is a good idea, I would suggest splitting this into smaller chunks so the reader has more information about what's going on when reading the tests body.

For example:

```go
func TestMyTest(t *testing.T) {
    tt := newReconcilerTest(t)
    tt.createAllObjs()
    
    logger := test.NewNullLogger()
	result, err := tt.reconciler().ReconcileControlPlane(tt.ctx, logger, tt.buildSpec())

	tt.Expect(err).NotTo(HaveOccurred())
	tt.Expect(tt.cluster.Status.FailureMessage).To(BeZero())
	tt.Expect(result).To(Equal(controller.Result{}))

    tt.checkCAPIClusterExists()
	tt.checkKubeadmControlPlaneExists()	
	tt.checkDockerMachineTemplateExists()
	tt.checkDockerClusterExists()
}
```

IMO this is almost as short as your tests, however, by adding a few repeated lines, this gives a way better idea of what the test is doing:
* I can see what steps the setup is doing, although abstracted in a couple methods
* I can clearly see the method that is being tested: `ReconcileControlPlane`
* I can clearly see what the expected returned values are
* I can see the additional expectations over the objects states and relate them to what the tested method is supposed to do

Reply:
Okay, I see your point. I've decided to leave the assertions in the test itself.